File: code/go/2021/apr/n-ary-tree-preorder-traversal.go
package apr

type Node struct {
	Val      int
	Children []*Node
}

func preorder(root *Node) []int {
	var result []int
	helper(root, &result)
	return result
}

func helper(root *Node, result *[]int) {
	if root != nil {
		*result = append(*result, root.Val)
		if root.Children != nil && len(root.Children) > 0 {
			for _, c := range root.Children {
				helper(c, result)
			}
		}
	}
}

File: codes/tree/n-ary-tree-preorder-traversal.cc
#include <vector>

using namespace std;

class Node
{
public:
    int val;
    vector<Node *> children;

    Node() {}

    Node(int _val)
    {
        val = _val;
    }

    Node(int _val, vector<Node *> _children)
    {
        val = _val;
        children = _children;
    }
};

class Solution
{
public:
    vector<int> preorder(Node *root)
    {
        vector<int> v;
        helper(&v, root);
        return v;
    }

    void helper(vector<int> *v, Node *node)
    {
        if (node)
        {
            for (auto n : node->children)
            {
                helper(v, n);
            }
            v->push_back(node->val);
        }
    }
};